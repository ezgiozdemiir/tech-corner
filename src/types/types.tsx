export interface CartItem extends Product {
    selectedColor: string;
    quantity: number;
  }
  
export interface CartState {
    cart: CartItem[];
    totalCost: number;
    customerBalance: number;
    addToCart: (product: Product, selectedColor: string, quantity: number) => void;
    removeFromCart: (productId: number) => void;
    clearCart: () => void;
}

export interface SelectDemoProps {
  colorOptions: string[];
  setedColor: (color: string) => void;
}

export interface Product {
  id: number;
  name: string;
  price: number;
  image: string;
  colorOptions: string[];
  type: string;
}
  
export interface SelectedColor {
  [productId: number]: string;
}
  
export interface Quantities {
  [productId: number]: number;
}
  
