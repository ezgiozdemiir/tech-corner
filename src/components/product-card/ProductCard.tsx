import { Button } from '../../components/ui/button'
import { ShoppingCart } from 'lucide-react'
import { SelectDemo } from '../../components/select/SelectDemo'
import { Product } from '@/types/types'
import React from 'react'

interface ProductCardProps {
    product: Product
    selectedColor: string
    setSelectedColor: (color: string) => void
    quantity: number
    setQuantity: (quantity: number) => void
    addToCart: (product: Product, color: string, quantity: number) => void
}

const ProductCard: React.FC<ProductCardProps> = ({
    product,
    selectedColor,
    setSelectedColor,
    quantity,
    setQuantity,
    addToCart,
}) => {
    return (
        <li className="product-card">
            <img src={product.image} alt={product.name} />
            <h2>{product.name}</h2>
            <p>Price: ${product.price}</p>
            <div className="color-selection">
                <div>Select Color:</div>
                <SelectDemo
                    colorOptions={product.colorOptions}
                    setedColor={setSelectedColor}
                />
            </div>
            <div style={{ marginTop: '10px' }}>
                <label>
                    Quantity:
                    <input
                        type="number"
                        value={quantity}
                        min="1"
                        onChange={(e) => setQuantity(Number(e.target.value))}
                        style={{
                            width: '50px',
                            marginLeft: '10px',
                        }}
                    />
                </label>
            </div>
            <Button
                className="button"
                onClick={() => addToCart(product, selectedColor, quantity)}
            >
                <ShoppingCart className="cart-button-icon" /> Add to Cart
            </Button>
        </li>
    )
}

export default ProductCard
