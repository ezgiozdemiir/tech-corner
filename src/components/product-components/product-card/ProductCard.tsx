import { Button } from '../../ui/button'
import { ShoppingCart } from 'lucide-react'
import { SelectDemo } from '../../select/SelectDemo'
import { Product } from '@/types/types'
import React, { forwardRef, useImperativeHandle, useRef } from 'react'
import { Link } from 'react-router-dom'
import { ProductCardProps } from '../interfaces/product.interface'

const ProductCard = forwardRef((props: ProductCardProps, ref) => {
    const inputRef = useRef<HTMLInputElement>(null)

    useImperativeHandle(ref, () => ({
        focusInput: () => {
            inputRef.current?.focus()
        },
    }))

    const {
        product,
        selectedColor,
        setSelectedColor,
        quantity,
        setQuantity,
        addToCart,
    } = props

    return (
        <li className="product-card">
            <Link to={`/products/id/${product.id}`}>
                <img
                    src={product.image}
                    alt={product.name}
                    className="product-image"
                />
            </Link>
            <h2>{product.name}</h2>
            <p>Price: ${product.price}</p>
            <div className="color-selection">
                <div>Select Color:</div>
                <SelectDemo
                    colorOptions={product.colorOptions}
                    setedColor={setSelectedColor}
                />
            </div>
            <div style={{ marginTop: '10px' }}>
                <label>
                    Quantity:
                    <input
                        ref={inputRef}
                        type="number"
                        value={quantity}
                        min="1"
                        onChange={(e) => setQuantity(Number(e.target.value))}
                        style={{
                            width: '50px',
                            marginLeft: '10px',
                        }}
                    />
                </label>
            </div>
            <button onClick={() => addToCart(product, selectedColor, quantity)}>
                <ShoppingCart className="cart-button-icon" /> Add to Cart
            </button>
        </li>
    )
})

export default ProductCard
