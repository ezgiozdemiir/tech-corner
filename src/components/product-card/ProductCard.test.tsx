import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import ProductCard from './ProductCard'
import { Product } from '@/types/types'
import userEvent from '@testing-library/user-event'

describe('ProductCard Component', () => {
    const mockProduct = {
        id: 1,
        name: 'Sample Product',
        price: 99.99,
        image: 'https://via.placeholder.com/150',
        colorOptions: ['Red', 'Blue', 'Green'],
        type: 'mobile',
    }

    const mockSetSelectedColor = jest.fn()
    const mockSetQuantity = jest.fn()
    const mockAddToCart = jest.fn()

    const defaultProps = {
        product: mockProduct,
        selectedColor: 'Red',
        setSelectedColor: mockSetSelectedColor,
        quantity: 1,
        setQuantity: mockSetQuantity,
        addToCart: mockAddToCart,
    }

    it('matches the snapshot', () => {
        const { asFragment } = render(<ProductCard {...defaultProps} />)
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders product details', () => {
        render(<ProductCard {...defaultProps} />)
        expect(screen.getByText('Sample Product')).toBeInTheDocument()
        expect(screen.getByText('Price: $99.99')).toBeInTheDocument()
    })

    it('calls setQuantity when quantity is changed', () => {
        render(<ProductCard {...defaultProps} />)
        const quantityInput = screen.getByLabelText('Quantity:')
        fireEvent.change(quantityInput, { target: { value: '3' } })
        expect(mockSetQuantity).toHaveBeenCalledWith(3)
    })

    it('calls addToCart when Add to Cart button is clicked', async () => {
        render(<ProductCard {...defaultProps} />)
        const button = screen.getByRole('button', { name: /add to cart/i })
        userEvent.click(button)
        await new Promise((resolve) => setTimeout(resolve, 100))
        expect(mockAddToCart).toHaveBeenCalledWith(mockProduct, 'Red', 1)
    })

    jest.mock('../../components/select/SelectDemo', () => ({
        SelectDemo: ({
            setedColor,
        }: {
            setedColor: (color: string) => void
        }) => (
            <div>
                <button onClick={() => setedColor('Blue')}>Blue</button>
                <button onClick={() => setedColor('Green')}>Green</button>
            </div>
        ),
    }))
})
