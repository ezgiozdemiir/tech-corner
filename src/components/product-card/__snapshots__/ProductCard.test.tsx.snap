// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Component matches the snapshot 1`] = `
<DocumentFragment>
  <li
    class="product-card"
  >
    <img
      alt="Sample Product"
      src="https://via.placeholder.com/150"
    />
    <h2>
      Sample Product
    </h2>
    <p>
      Price: $99.99
    </p>
    <div
      class="color-selection"
    >
      <div>
        Select Color:
      </div>
      <button
        aria-autocomplete="none"
        aria-controls="radix-:r0:"
        aria-expanded="false"
        class="flex h-9 w-full items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 select-trigger"
        data-placeholder=""
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          Select a color
        </span>
        <svg
          aria-hidden="true"
          class="h-4 w-4 opacity-50"
          fill="none"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
    <div
      style="margin-top: 10px;"
    >
      <label>
        Quantity:
        <input
          min="1"
          style="width: 50px; margin-left: 10px;"
          type="number"
          value="1"
        />
      </label>
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 button"
    >
      <svg
        class="lucide lucide-shopping-cart cart-button-icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="8"
          cy="21"
          r="1"
        />
        <circle
          cx="19"
          cy="21"
          r="1"
        />
        <path
          d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"
        />
      </svg>
       Add to Cart
    </button>
  </li>
</DocumentFragment>
`;
